src/controllers/user.controller.ts

```diff
@@ -11,15 +11,34 @@ import {hash} from 'bcryptjs';
 import {promisify} from 'util';
 import * as isemail from 'isemail';
 import {RecommenderService} from '../services/recommender.service';
-import {inject} from '@loopback/core';
+import {inject, Setter} from '@loopback/core';
+import {
+  authenticate,
+  UserProfile,
+  AuthenticationBindings,
+} from '@loopback/authentication';
+import {Credentials} from '../repositories/user.repository';
 
 const hashAsync = promisify(hash);
 
+// TODO(jannyHou): This should be moved to @loopback/authentication
+const UserProfileSchema = {
```

Comment:
We need to find a way allowing applications to add additional properties to user profile.

I would also prefer to leverage `@model` and `@property` decorators together with the existing code converting LB4 models to JSON/OpenAPI schema, instead of maintaining the schema manually.

```ts
@model
class UserProfile extends Model {
  @property({type: 'string', required: true})
  id: string;

  @property()
  email: string;

  @property()
  name: string;
}
```

Feel free to leave such changes out of scope of this spike and create a new user story to follow up on this area.

Reply:
There will be a refactor PR in `@loopback/authentication` coming soon, this change will definitely be part of it :)